Make Dask configurable in configuration (#2616)
ESMValGroup/ESMValCore: doc/quickstart/configure.rst

Lazy data is represented by `dask arrays <https://docs.dask.org/en/stable/array.html>`__.
Dask arrays consist of many small
`numpy arrays <https://numpy.org/doc/stable/user/absolute_beginners.html#what-is-an-array>`__
(called chunks) and if possible, computations are run on those small arrays in
parallel.
In order to figure out what needs to be computed when, Dask makes use of a
'`scheduler <https://docs.dask.org/en/stable/scheduling.html>`__'.
The default scheduler in Dask is rather basic, so it can only run on a single
computer and it may not always find the optimal task scheduling solution,
resulting in excessive memory use when using e.g. the
:func:`esmvalcore.preprocessor.multi_model_statistics` preprocessor function.
Therefore it is recommended that you take a moment to configure the
`Dask distributed <https://distributed.dask.org>`__ scheduler.
A Dask scheduler and the 'workers' running the actual computations, are
collectively called a 'Dask cluster'.

Predefined clusters
--------------------

ESMValCore provides a set of predefined clusters, which can be selected in a
YAML configuration file via

.. code:: yaml

  dask:

>     run: <NAME_OF_CLUSTER>

Author: Done in https://github.com/ESMValGroup/ESMValCore/pull/2616/commits/f9f4315add2d342b821a6b8818c900824cf81f49

    clusters:
      local_cluster:
        type: distributed.LocalCluster

This should work well for most personal computers.

.. note::

   If running this configuration on a shared node of an HPC cluster, Dask will
   try and use as many resources it can find available, and this may lead to
   overcrowding the node by a single user (you)!

*Shared computer*

Create a :class:`distributed.LocalCluster` on the computer running ESMValCore,
with 2 workers with 2 threads/4 GiB of memory each (8 GiB in total):

.. code:: yaml

  dask:
    run: local_cluster  # use "local_cluster" defined below
    clusters:
      local_cluster:
        type: distributed.LocalCluster
        n_workers: 2
        threads_per_worker: 2
        memory_limit: 4GiB

this should work well for shared computers.

*Computer cluster*

Create a Dask distributed cluster on the `Levante
<https://docs.dkrz.de/doc/levante/running-jobs/index.html>`__ supercomputer
using the `Dask-Jobqueue <https://jobqueue.dask.org/en/latest/>`__ package:

.. code:: yaml

  dask:
    run: slurm_cluster  # use "slurm_cluster" defined below
    clusters:
      slurm_cluster:
        type: dask_jobqueue.SLURMCluster
        queue: shared

>         account: bk1088

Reviewer: I'd remove this as public info, Manu

Author: Done in https://github.com/ESMValGroup/ESMValCore/pull/2616/commits/a40fb69124649952bdc45b535d6269429658f5ee

        cores: 8
        memory: 7680MiB
        processes: 2
        interface: ib0

>         local_directory: "/scratch/b/b381141/dask-tmp"

Reviewer: valid user name, pls remove this as public information

Author: Done in https://github.com/ESMValGroup/ESMValCore/pull/2616/commits/a40fb69124649952bdc45b535d6269429658f5ee

   will be downloaded; otherwise, local data will be used.

.. _config-dask:

Dask configuration
==================

Configure Dask in the  ``dask`` section.

The :ref:`preprocessor functions <preprocessor_functions>` and many of the
:ref:`Python diagnostics in ESMValTool <esmvaltool:recipes>` make use of the
:ref:`Iris <iris:iris_docs>` library to work with the data.
In Iris, data can be either :ref:`real or lazy <iris:real_and_lazy_data>`.
Lazy data is represented by `dask arrays <https://docs.dask.org/en/stable/array.html>`__.
Dask arrays consist of many small
`numpy arrays <https://numpy.org/doc/stable/user/absolute_beginners.html#what-is-an-array>`__
(called chunks) and if possible, computations are run on those small arrays in
parallel.
In order to figure out what needs to be computed when, Dask makes use of a
'`scheduler <https://docs.dask.org/en/stable/scheduling.html>`__'.

> The default scheduler in Dask is rather basic, so it can only run on a single

Reviewer 2: ```suggestion
The thread-based scheduler in Dask is rather basic, so it can only run on a single
```

computer and it may not always find the optimal task scheduling solution,
resulting in excessive memory use when using e.g. the
:func:`esmvalcore.preprocessor.multi_model_statistics` preprocessor function.
Therefore it is recommended that you take a moment to configure the
`Dask distributed <https://distributed.dask.org>`__ scheduler.
A Dask scheduler and the 'workers' running the actual computations, are
collectively called a 'Dask cluster'.

> Predefined Dask profiles

Reviewer 2: ```suggestion
Dask profiles
```

Author: Done in https://github.com/ESMValGroup/ESMValCore/pull/2616/commits/a5368277069ff0cc29caafbec5ce0b864b84fbd6

------------------------

> ESMValCore provides a set of predefined Dask profiles, which can be selected in

Reviewer 2: ```suggestion
Because some recipes require more computational resources than others, ESMValCore provides the option to define "Dask profiles". These profiles can be used to update the `Dask user configuration <https://docs.dask.org/en/stable/configuration.html>`__ per recipe run. The Dask profile can selected in
```

a YAML configuration file via

.. code:: yaml

  dask:
    use: <NAME_OF_PROFILE>

or alternatively in the command line via

.. code:: bash

  esmvaltool run --dask='{"use": "<NAME_OF_PROFILE>"}' recipe_example.yml

Available predefined Dask profiles:

> - ``local_threaded`` (selected by default): use `default thread-based scheduler

Reviewer 2: ```suggestion
- ``local_threaded`` (selected by default): use `thread-based scheduler
```

  <https://docs.dask.org/en/stable/scheduling.html#local-threads>`__ without
  any further options.
- ``local_distributed``: use `local distributed scheduler
  <https://docs.dask.org/en/stable/scheduling.html#dask-distributed-local>`__
  without any further options.
- ``debug``: use `synchronous Dask scheduler
  <https://docs.dask.org/en/stable/scheduling.html#single-thread>`__ for
  debugging purposes.
  Best used with ``max_parallel_tasks: 1``.

> Custom Dask distributed scheduler configuration

Reviewer 2: ```suggestion
Dask distributed scheduler configuration
```

Author: Done in https://github.com/ESMValGroup/ESMValCore/pull/2616/commits/a5368277069ff0cc29caafbec5ce0b864b84fbd6

.. code:: yaml

  dask:
    use: <NAME_OF_CLUSTER>

or alternatively with in command line via

.. code:: bash

>   esmvaltool run --dask='{"use": "<NAME_OF_CLUSTER>"}' recipe_example.yml

Reviewer 2: This syntax is a bit awkward. Do you think it would be possible to use the syntax where you can provide nested dicts using dots, as provided by `dask.config`, i.e. `esmvaltool run --dask.use = <NAME_OF_CLUSTER>`?

Reviewer 2: If not, maybe it would be nice to just add an extra command line argument to specify the profile, e.g. `esmvaltool run --dask-profile local-distributed recipe_python.yml`

Author: Unfortunately, using something like `--dask.use` is not possible at the moment. The logic that handles this properly for Dask is (I believe) implemented in `dask.config.set`, which we do not use.

I agree that the existing syntax is a bit clunky, but I really don't want to add another configuration option. This will be really confusing.

How about we try to make the syntax `--dask.use` work in a separate PR? This would also come in handy for other configuration options that use nested dictionaries (which I think we will use more in the future).

Reviewer 2: Sounds good

.. _config-dask-default-scheduler:

Custom Dask default scheduler configuration
-------------------------------------------

The Dask default scheduler can be a good choice for recipes using a small
amount of data or when running a recipe where not all preprocessor functions
are lazy yet (see :issue:`674` for the current status).

To avoid running out of memory, it is important to set the number of workers
(threads) used by Dask to run its computations to a reasonable number.
By default the number of CPU cores in the machine will be used, but this may be
too many on shared machines or laptops with a large number of CPU cores
compared to the amount of memory they have available.

Typically, Dask requires about 2 GiB of RAM per worker, but this may be more
depending on the computation.

To set the number of workers used by the Dask default scheduler, use the
following configuration:

.. code:: yaml

  dask:
    run: default  # This can be omitted
    clusters:
      default:

>         type: default

Author: Renamed to `local_threaded` in https://github.com/ESMValGroup/ESMValCore/pull/2616/commits/a40fb69124649952bdc45b535d6269429658f5ee

.. _config-dask-default-scheduler:

> Custom Dask default scheduler configuration

Reviewer 2: ```suggestion
Dask threaded scheduler configuration
```
To make it more specific?

Author: Done in https://github.com/ESMValGroup/ESMValCore/pull/2616/commits/f6fc37cf23c291066bc2f6fc6015860856c83c3c

-------------------------------------------

> The Dask default scheduler can be a good choice for recipes using a small

Reviewer 2: ```suggestion
The Dask threaded scheduler can be a good choice for recipes using a small
```

Reviewer 2: It would be nice to reduce the use of the word "default" since it does not convey much meaning and means different things in different contexts.

amount of data or when running a recipe where not all preprocessor functions
are lazy yet (see :issue:`674` for the current status).

To avoid running out of memory, it is important to set the number of workers
(threads) used by Dask to run its computations to a reasonable number.
By default, the number of CPU cores in the machine will be used, but this may
be too many on shared machines or laptops with a large number of CPU cores
compared to the amount of memory they have available.

Typically, Dask requires about 2 GiB of RAM per worker, but this may be more
depending on the computation.

> To set the number of workers used by the (thread-based) Dask default scheduler,

Reviewer 2: ```suggestion
To set the number of workers used by the threaded scheduler,
```

Author: Done in https://github.com/ESMValGroup/ESMValCore/pull/2616/commits/f6fc37cf23c291066bc2f6fc6015860856c83c3c

use the following configuration:

.. code:: yaml

  dask:
    use: local_threaded  # This can be omitted
    profiles:
      local_threaded:
        num_workers: 4

To switch to a process-based scheduler, use:

.. code:: yaml

  dask:
    use: local_processes
    profiles:
      local_processes:
        num_workers: 4

>         scheduler: processes

Reviewer 2: This scheduler is [not supported by Iris](https://github.com/SciTools/iris/blob/ec447312cbd05b1a599118efa5d829e5dfaeb3e1/lib/iris/fileformats/netcdf/_dask_locks.py#L46), so I would leave this out.

Author: Removed in https://github.com/ESMValGroup/ESMValCore/pull/2616/commits/28d2ee09a2e73aadf55e0dd64c6892c2cf422beb
